dsgrid/project.py

```diff
@@ -183,32 +170,26 @@ def process_query(self, context: QueryContext, cached_project_mapped_datasets_di
         dim_columns = sorted(dim_columns)
         time_columns = sorted(time_columns)
         expected_columns = time_columns + pivoted_columns_sorted + dim_columns
-        for i, df in enumerate(dfs):
+
+        datasets = {}
+        for dataset_id, path in df_filenames.items():
+            df = read_dataframe(path)
             remaining = sorted(set(df.columns).difference(expected_columns))
             final_columns = expected_columns + remaining
             missing = pivoted_columns.difference(df.columns)
             for column in missing:
                 df = df.withColumn(column, F.lit(None).cast(DoubleType()))
-            df = df.select(*final_columns)
-            dfs[i] = df
-
-        main_df = dfs[0]
-        if len(dfs) > 1:
-            for df in dfs[1:]:
-                if df.columns != main_df.columns:
-                    raise Exception(
-                        f"BUG: dataset columns must match. main={main_df.columns} new={df.columns} "
-                        f"diff={set(main_df.columns).symmetric_difference(df.columns)}"
-                    )
-                main_df = main_df.union(df)
+            datasets[dataset_id] = DatasetExpressionHandler(
+                df.select(*final_columns), time_columns + dim_columns, pivoted_columns_sorted
+            )
 
-        return main_df
+        return evaluate_expression(context.model.project.dataset.expression, datasets).df
```

Comment:
I think we need to do `df.select(*final_columns)` here after `evaluate_expression()`

Reply:
They have to be in the correct order before evaluate_expression; otherwise the result is garbage.

Reply:
I think this is a stale comment now since `final_columns` is not defined and not being enforced here.